sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/SsrfInjection.kt

```diff
@@ -0,0 +1,157 @@
+ * ranging from information disclosure to exposure of internal services.
+ *
+ * The sanitizer detects such attacks by checking if the fuzzer was able to inject the honeypot host name
+ * into the requested destination. If this is the case, it reports a high severity security issue.
+ *
+ * That being said, if fuzzer input reaches this detector, it's likely to already trigger "slow input"
+ * findings, due to trying to reach invalid hosts, before the honeypot host is requested.
+ *
+ * Checks are implemented via [java.net.Socket.connect] and [java.nio.channels.SocketChannel.connect] methods.
+ */
+@Suppress("unused_parameter", "unused")
+object SsrfInjection {
+
+    /**
+     * Honeypot host name targeting an invalid address.
+     * RFC 2606 defines such names at https://www.rfc-editor.org/rfc/rfc2606#section-2
+     */
+    private const val HONEYPOT_HOST: String = "jazzer.invalid"
+
+    // [java.nio.channels.SocketChannel] is used in many JDK classes to open
+    // (non-blocking) network connections. E.g. [java.net.http.HttpClient] uses
+    // it internally.
+    //
+    // The actual connection is established in the abstract "connect" method.
+    // Hooking that also hooks all concrete implementations, from which
+    // only one exists in [sun.nio.ch.SocketChannelImpl]. "connect" is only
+    // called in [java.nio.channels.SocketChannel] itself and the two
+    // mentioned classes.
+    @MethodHook(
+        type = HookType.BEFORE,
+        targetClassName = "java.nio.channels.SocketChannel",
+        targetMethod = "connect",
+        additionalClassesToHook = [
+            "sun.nio.ch.SocketAdaptor",
+            "jdk.internal.net.http.PlainHttpConnection",
+        ],
+    )
+    @JvmStatic
+    fun checkSsrfHttpConnection(method: MethodHandle, thisObject: Any?, arguments: Array<Any>, hookId: Int) {
+        checkSsrf(arguments, hookId)
+    }
+
+    // [java.net.Socket] is used in many JDK classes to open network connections.
+    // Internally it delegates to [java.net.SocketImpl], hence, for most situations
+    // it's sufficient to hook [java.net.Socket] itself.
+    //
+    // As [java.net.SocketImpl] is an abstract class, so all concrete implementations
+    // get hooked. Internal classes invoking connect on those implementations need
+    // to be specified explicitly, though.
+    //
+    // Internal classes "using" [java.net.SocketImpl.connect]:
+    // - java.net.Socket (hook required)
+    // - java.net.AbstractPlainSocketImpl (no direct usage)
+    // - java.net.PlainSocketImpl (no direct usage)
+    // - java.net.HttpConnectSocketImpl (only used in Socket)
+    // - java.net.SocksSocketImpl (used in Socket, but also invoking super.connect directly, hook required)
+    // - java.net.ServerSocket (security check, no hook required)
+    @MethodHook(
+        type = HookType.BEFORE,
+        targetClassName = "java.net.SocketImpl",
+        targetMethod = "connect",
+        additionalClassesToHook = [
```

Comment:
I find the comment with the `additionalClassesToHook` difficult to understand at first glance. Why do we need to hook the 4 additional classes, from which only 2 appear in the above comment? What about those that are explicitly mentioned in the comment?

Reply:
I tried to explain the reason to hook those classes in more detail. The classes that were not mentioned are not required, though. Thx for noticing that!